mkosi/machine.py

```diff
@@ -122,9 +143,52 @@ def run(self, commands: Sequence[str], timeout: int = 900, check: bool = True) -
 
         return process
 
-    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+    def _kill(self, *args: Any, **kwargs: Any) -> None:
         if self._serial:
             self._serial.kill(signal.SIGTERM)
             self.exit_code = self._serial.wait()
             self._serial = None
         self.stack.__exit__(*args, **kwargs)
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        self._kill(*args, **kwargs)
+
+    def reboot(self) -> None:
+        self._kill(None, None, None)
+        self._boot()
+
+
+class MkosiMachineTest(unittest.TestCase):
+    args: Sequence[str]
+    machine: Machine
+    exc_info: Tuple[Optional[Type[BaseException]], Optional[BaseException], Optional[TracebackType]] \
+              = (None, None, None)
+
+    def _callTestMethod(self, method: str) -> None:
+        try:
+            super()._callTestMethod(method) # type: ignore
+        except Exception as e:
+            self.exc_info = exc_info()
+            raise e
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.machine = Machine(cls.args)
+
+        machine_verb = cls.machine.args.verb
+        no_nspawn = parse_boolean(os.getenv("MKOSI_TESTS_NO_NSPAWN", "0"))
+        no_qemu = parse_boolean(os.getenv("MKOSI_TESTS_NO_QEMU", "0"))
+
+        if no_nspawn and machine_verb == Verb.boot:
+            raise unittest.SkipTest("Nspawn test skipped due to environment variable.")
+        if no_qemu and machine_verb == Verb.qemu:
+            raise unittest.SkipTest("Qemu test skipped due to environment variable.")
+
+        cls.machine.__enter__()
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        cls.machine.__exit__(*cls.exc_info)
+
+    def tearDown(self) -> None:
+        self.machine.reboot()
```

Comment:
Hmm, so one realization I just had here is that calling reboot() in tearDown() means we'll unnecessarily boot the machine after the last test.

So maybe calling reboot() here was the wrong call from me, my bad. What we can do instead is make boot() and kill() public methods of the Machine class, and call kill() in tearDown() and boot() in setUp(). That's equivalent to rebooting but will avoid booting the machine unnecessarily after the last test finishes.